File: lib/models/visitor.dart
/// A visitor logged against a property — [exitTime] stays null until
/// they're checked out, which is what distinguishes "currently inside"
/// from history.
class Visitor {
  final String id;
  final String name;
  final String? phone;
  final String? purpose;
  final String? meetingName;
  final DateTime entryTime;
  final DateTime? exitTime;

  const Visitor({
    required this.id,
    required this.name,
    this.phone,
    this.purpose,
    this.meetingName,
    required this.entryTime,
    this.exitTime,
  });

  bool get isInside => exitTime == null;

  factory Visitor.fromJson(Map<String, dynamic> json) => Visitor(
        id: json['_id'] as String,
        name: json['name'] as String,
        phone: json['phone'] as String?,
        purpose: json['purpose'] as String?,
        meetingName: json['meetingName'] as String?,
        entryTime: DateTime.parse(json['entryTime'] as String),
        exitTime: json['exitTime'] != null ? DateTime.parse(json['exitTime'] as String) : null,
      );
}

File: backend/models/Visitor.js
const mongoose = require('mongoose');

// A visitor logged against a property - entryTime is set the moment the
// owner logs them in; exitTime stays null until they're checked out, which
// is what distinguishes "currently inside" from history.
const visitorSchema = new mongoose.Schema(
  {
    property: { type: mongoose.Schema.Types.ObjectId, ref: 'Property', required: true, index: true },
    owner: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    name: { type: String, required: true, trim: true },
    phone: { type: String, trim: true },
    purpose: { type: String, trim: true },
    // Who they're visiting - free text (tenant name/room), not a hard
    // reference, since a visitor might be here for the owner or nobody
    // specific (e.g. a delivery, a technician).
    meetingName: { type: String, trim: true },
    entryTime: { type: Date, default: Date.now },
    exitTime: { type: Date, default: null },
  },
  { timestamps: true }
);

module.exports = mongoose.model('Visitor', visitorSchema);
